### DISCUSSION TITLE
How to make Custom Cue Points work, without enable AutoCue

### DISCUSSION THREAD

**Discussion by unudelaradio** [AzuraCast/AzuraCast#7272] [Q&A]
Hello guys!

I'm desperately asking for your help, and i hope @RM-FM and @Moonbase59 could give me some advice. After updating my Azuracast instance from v0.19.7 to the latest Rolling Release, i'm facing some issues and i have some questions:

First, all my files have custom Cue Points edited in Visual Cue Editor and i'm not using AutoCue.

1. On v0.19.7 i was using this custom script made by RM-FM, placed in the 3rd box of Liquidsoap Config:

```
def clean_transition(old, new) =

  def safe_float_of_string(str) =
    try
      float_of_string(str)
    catch _ do
      0.
    end
  end

  cross_duration = safe_float_of_string(old.metadata["liq_cross_duration"])
  fade_out = safe_float_of_string(old.metadata["liq_fade_out"])
  fade_in = safe_float_of_string(new.metadata["liq_fade_in"])
  log_label = "clean_transition"
    
  let fade.out = fun (s) -> fade.out(type="exp", duration=fade_out, s)
  let fade.in = fun (s) -> fade.in(type="log", duration=fade_in, s)

  add = fun (a, b) -> add(normalize=false, [b, a])
  sequence = fun (a, b) -> (sequence([a, b]) : source)

  if cross_duration > 0. then
    log(label= log_label,"Crossfade. Duration: #{cross_duration}. Fade out: #{fade_out}. Fade in: #{fade_in}.")
    if fade_out > 0. and fade_in > 0. then 
      add(fade.out(old.source), fade.in(new.source))
    elsif fade_out > 0. and fade_in <= 0. then 
      add(fade.out(old.source), new.source)
    elsif fade_out <= 0. and fade_in > 0. then 
      add(old.source, fade.in(new.source))
    else
      add(old.source, new.source)
    end
  else
    log(label= log_label,"Sequence. Fade out: #{fade_out}. Fade in: #{fade_in}.")
    if fade_out > 0. and fade_in > 0. then 
      sequence(fade.out(old.source), fade.in(new.source))
    elsif fade_out > 0. and fade_in <= 0. then 
      sequence(fade.out(old.source), new.source)
    elsif fade_out <= 0. and fade_in > 0. then 
      sequence(old.source, fade.in(new.source))
    else
      sequence(old.source, new.source)
    end
  end

end

radio = cross(clean_transition,radio)
```
Should this script work on newest AzuraCast version? Because all my transitions are broken now with lots of blanks and strange fades.

2. Because the custom script above doesn't work anymore, i cleaned all my Liquidsoap Config and tried the basic AzuraCast method: _Crossfade Method > Normal Mode > 0.1 Crossfade Duration_ in the AutoDJ settings. Still my Cue Points don't work and i have blanks after every track.

3. After trying the 2 methods above, i thouth my Cue Poins are lost or missing. So i took a look in Visual Cue Editor. I discovered that the visual waveforms are also broken (only a zoom shows the correct visual representation). In the Advanced tab of Edit Media, the custom values are present. But still, there are some issues: **when exporting the CSV file from Bulk Media Import/Export, all the custom Cue Points values are empty.**

So the final question is: how to achieve the functionality from point 1 in this newest AzuraCast version without activate AutoCue? Maybe someone could help me with a custom LS config script, similar with the one i was using in v0.19.7. I just want my Cue Points to work as before. 😔

**Comment by unudelaradio** [comment]
After another 24 hours of tests, i'm still trying to understand how can i make my Custom Cue Points work, without enable AutoCue. All my tracks are carefully edited in Visual Cue Editor. Do Custom Cue Points work with any Crossfade Method in AutoDJ, in newest AzuraCast version?

**Comment by RM-FM** [comment]
> Hello guys!
> 
> I'm desperately asking for your help, and i hope @RM-FM and @Moonbase59 could give me some advice. After updating my Azuracast instance from v0.19.7 to the latest Rolling Release, i'm facing some issues and i have some questions:
> 
> First, all my files have custom Cue Points edited in Visual Cue Editor and i'm not using AutoCue.
> 
> 1. On v0.19.7 i was using this custom script made by RM-FM, placed in the 3rd box of Liquidsoap Config:
> 
> ```
> def clean_transition(old, new) =
> 
>   def safe_float_of_string(str) =
>     try
>       float_of_string(str)
>     catch _ do
>       0.
>     end
>   end
> 
>   cross_duration = safe_float_of_string(old.metadata["liq_cross_duration"])
>   fade_out = safe_float_of_string(old.metadata["liq_fade_out"])
>   fade_in = safe_float_of_string(new.metadata["liq_fade_in"])
>   log_label = "clean_transition"
>     
>   let fade.out = fun (s) -> fade.out(type="exp", duration=fade_out, s)
>   let fade.in = fun (s) -> fade.in(type="log", duration=fade_in, s)
> 
>   add = fun (a, b) -> add(normalize=false, [b, a])
>   sequence = fun (a, b) -> (sequence([a, b]) : source)
> 
>   if cross_duration > 0. then
>     log(label= log_label,"Crossfade. Duration: #{cross_duration}. Fade out: #{fade_out}. Fade in: #{fade_in}.")
>     if fade_out > 0. and fade_in > 0. then 
>       add(fade.out(old.source), fade.in(new.source))
>     elsif fade_out > 0. and fade_in <= 0. then 
>       add(fade.out(old.source), new.source)
>     elsif fade_out <= 0. and fade_in > 0. then 
>       add(old.source, fade.in(new.source))
>     else
>       add(old.source, new.source)
>     end
>   else
>     log(label= log_label,"Sequence. Fade out: #{fade_out}. Fade in: #{fade_in}.")
>     if fade_out > 0. and fade_in > 0. then 
>       sequence(fade.out(old.source), fade.in(new.source))
>     elsif fade_out > 0. and fade_in <= 0. then 
>       sequence(fade.out(old.source), new.source)
>     elsif fade_out <= 0. and fade_in > 0. then 
>       sequence(old.source, fade.in(new.source))
>     else
>       sequence(old.source, new.source)
>     end
>   end
> 
> end
> 
> radio = cross(clean_transition,radio)
> ```
> 
> Should this script work on newest AzuraCast version? Because all my transitions are broken now with lots of blanks and strange fades.
> 
> 2. Because the custom script above doesn't work anymore, i cleaned all my Liquidsoap Config and tried the basic AzuraCast method: _Crossfade Method > Normal Mode > 0.1 Crossfade Duration_ in the AutoDJ settings. Still my Cue Points don't work and i have blanks after every track.
> 3. After trying the 2 methods above, i thouth my Cue Poins are lost or missing. So i took a look in Visual Cue Editor. I discovered that the visual waveforms are also broken (only a zoom shows the correct visual representation). In the Advanced tab of Edit Media, the custom values are present. But still, there are some issues: **when exporting the CSV file from Bulk Media Import/Export, all the custom Cue Points values are empty.**
> 
> So the final question is: how to achieve the functionality from point 1 in this newest AzuraCast version without activate AutoCue? Maybe someone could help me with a custom LS config script, similar with the one i was using in v0.19.7. I just want my Cue Points to work as before. 😔

This script won't work anymore due to changes in Liquidsoap meta data related to crossfades. Please remove the custom code and enable the Azuracast native crossfade feature. There were plenty of improvements related to crossfades. I don't think that you need custom LS code anymore for clean transitions.

**Reply by unudelaradio** [reply]
✅ All my LS Config boxes are cleaned.
✅ All my Custom Cue Points are carefully edited. 
✅ `Crossfade Method > Normal > 0.1 Seconds`

Example of how my Cue Points look (this Jingle is played end-to-end... with no overlap):

![Screenshot 2024-07-03 105450](https://github.com/AzuraCast/AzuraCast/assets/74244061/2f22b3ea-610f-4e41-bf23-43281831d7d4)

What am i doing wrong?

**Reply by unudelaradio** [reply]
I have a feeling that Crossfade Duration (Seconds) in AutoDJ settings, overrides custom cue points.

Custom `liq_cross_start_next` and `liq_fade_out` shouldn't be prioritized?

So when i edit my Next Start and Fade Out points in Media, the default 0.1 value in AutoDJ settings should be ignored. The default `0.1` should be applied only if specific Media fields are empty.

Now it seems that AutoDJ default setting is prioritized.

**Comment by Moonbase59** [comment]
Working on it, see https://github.com/AzuraCast/AzuraCast/issues/7281
